addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java

```diff
@@ -32,55 +31,69 @@
-import org.eclipse.smarthome.core.thing.type.ChannelTypeUID;
+import org.eclipse.smarthome.core.thing.type.ChannelTypeProvider;
 import org.eclipse.smarthome.core.types.Command;
 import org.eclipse.smarthome.core.types.RefreshType;
 import org.openhab.binding.yamahareceiver.YamahaReceiverBindingConstants;
-import org.openhab.binding.yamahareceiver.internal.YamahaChannelsTypeProvider;
-import org.openhab.binding.yamahareceiver.internal.protocol.InputWithNavigationControl;
-import org.openhab.binding.yamahareceiver.internal.protocol.InputWithPlayControl;
-import org.openhab.binding.yamahareceiver.internal.protocol.InputWithPlayControl.PlayControlState;
-import org.openhab.binding.yamahareceiver.internal.protocol.InputWithPlayControl.PlayInfoState;
-import org.openhab.binding.yamahareceiver.internal.protocol.ZoneControl;
-import org.openhab.binding.yamahareceiver.internal.protocol.ZoneControl.AvailableInputState;
-import org.openhab.binding.yamahareceiver.internal.protocol.ZoneControl.Listener;
-import org.openhab.binding.yamahareceiver.internal.protocol.ZoneControl.State;
-import org.openhab.binding.yamahareceiver.internal.protocol.ZoneControl.Zone;
+import org.openhab.binding.yamahareceiver.internal.ChannelsTypeProviderAvailableInputs;
+import org.openhab.binding.yamahareceiver.internal.ChannelsTypeProviderPreset;
+import org.openhab.binding.yamahareceiver.internal.protocol.*;
+import org.openhab.binding.yamahareceiver.internal.protocol.AbstractConnection;
+import org.openhab.binding.yamahareceiver.internal.protocol.xml.InputWithNavigationControlXML;
+import org.openhab.binding.yamahareceiver.internal.protocol.xml.InputWithPlayControlXML;
+import org.openhab.binding.yamahareceiver.internal.protocol.xml.ZoneControlXML;
+import org.openhab.binding.yamahareceiver.internal.state.*;
+import org.osgi.framework.ServiceRegistration;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.xml.sax.SAXException;
 
 /**
  * The {@link YamahaZoneThingHandler} is managing one zone of an Yamaha AVR.
- * It has a state consisting of the zone, the current input ID, {@link ZoneControl.State}
+ * It has a state consisting of the zone, the current input ID, {@link ZoneControlState}
  * and some more state objects and uses the zone control protocol
- * class {@link ZoneControl}, {@link InputWithPlayControl} and {@link InputWithNavigationControl}
+ * class {@link ZoneControlXML}, {@link InputWithPlayControlXML} and {@link InputWithNavigationControlXML}
  * for communication.
  *
  * @author David Graeff <david.graeff@web.de>
+ * @author Tomasz Maruszak - [yamaha] Tuner band selection and preset feature for dual band models (RX-S601D)
  */
-public class YamahaZoneThingHandler extends BaseThingHandler
-        implements Listener, InputWithNavigationControl.Listener, InputWithPlayControl.Listener {
+public class YamahaZoneThingHandler extends BaseThingHandler implements ZoneControlStateListener,
```

Comment:
True, but this is enterprise ready software, it needs a hell lot of interfaces and factories :rofl: